### PR TITLE
when building a layered docker image, ignore it if tar encounters cha…

### PR DESCRIPTION
**Pull request by purefn**
###### Motivation for this change
Allows building layered docker images in parallel with other derivations. Takes the second approach mentioned in #75888, telling tar to ignore the changes to `/nix/store`. `tar` exit codes are defined as

```
       0      Successful termination.

       1      Some files differ.  If tar was invoked with the --compare (--diff, -d) command line option, this means that some files in the  archive  differ
              from  their  disk counterparts.  If tar was given one of the --create, --append or --update options, this exit code means that some files were
              changed while being archived and so the resulting archive does not contain the exact copy of the file set.

       2      Fatal error.  This means that some fatal, unrecoverable error occurred.
```
So as long as `tar` exits with a `0` or `1` we should be fine. The rationale for that is that any derivations that building the layered docker image depends on would have already been built and in the nix store by the time we get to this point. Therefore, any changes to `/nix/store` don't pertain to the docker image under construction.

###### Things done

- [x] Tested using sandboxing ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nix-review --run "nix-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Ensured that relevant documentation is up to date
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

###### Notify maintainers

cc @nlewo

### CODE REVIEW

**Comment by jonringer** [comment]
cc @grahamc

**Comment by purefn** [comment]
Hmm I'm still seeing an intermittent error 

```
docker-image-learning-server-nix.tar.gz> Cooking the image...
docker-image-learning-server-nix.tar.gz>  tar: ./6773e5a7a063b5386300f1675b617904d21ee4b8c7f9d75b275c67a62cf18e0a/layer.tar: file changed as we read it
```

That seems to be coming from this bit

```
echo "Cooking the image..."                                                                                                                        
tar -C image --dereference --hard-dereference --sort=name --mtime="@$SOURCE_DATE_EPOCH" --owner=0 --group=0  --mode=a-w --xform s:'^./':: -c . |       pigz -nT > $out       
```

I'm a bit confused. I don't see how that file could be modified at this point and I'm hesitant to give it the same treatment without better understanding.

**Comment by nlewo** [comment]
I'm not in favor of this approach. I would prefer to explore the approach described in https://github.com/NixOS/nixpkgs/issues/75888#issuecomment-567833985.

**Comment by purefn** [comment]
@nlewo - I've updated it with a different approach which should work better. Let me know what you think!

File: pkgs/build-support/docker/store-path-to-layer.sh [hunk 1]
```diff
@@ -5,16 +5,29 @@ set -eu
...
+    --transform='s,nix,/nix,' \
+    nix
+
+n=$(basename "$storePath")
+tar -C /nix/store -rpf "$layerPath/layer.tar" \
+    --hard-dereference --sort=name \
     --mtime="@$SOURCE_DATE_EPOCH" \
-    --owner=0 --group=0 "$@"
+    --owner=0 --group=0 \
+    --transform="s,$n,/nix/store/$n," \
```
Comment (nlewo) [inline pkgs/build-support/docker/store-path-to-layer.sh]: I don't understand why this change would be required. If I remember well, tar doesn't create entries for the directories `/nix` and `/nix/store` in the tar archive. Since they are not explicitly created by tar, it doesn't read them and I don't see any issues here regarding these entries.

**Review by nlewo — CHANGES_REQUESTED** [review]
Otherwise, your first change looks fine to me ;)

File: pkgs/build-support/docker/store-path-to-layer.sh (hunk 1 above)
Comment (purefn) [reply pkgs/build-support/docker/store-path-to-layer.sh]: You are correct that `tar` does not create entries for `/nix` and `/nix/store`. But even though it doesn't create those entries, it does still check if they are modified while it's running. I was able to force `tar` to fail in this way by running a `nix build` of a bunch of packages and at the same time running 

```
tar -cf test.tar /nix/store/vs7zapqwngylvpydpv5dg3gk1va8dj8w-bash-interactive-4.4-p23
```

repeatedly.

Comment (nlewo) [reply pkgs/build-support/docker/store-path-to-layer.sh]: Ok. Nice stress test and thx for your explanation.

File: pkgs/build-support/docker/store-path-to-layer.sh [hunk 2]
```diff
@@ -5,16 +5,29 @@ set -eu
...
+shift
+
 layerPath="./layers/$layerNumber"
-echo "Creating layer #$layerNumber for $@"
+echo "Creating layer #$layerNumber for $storePath"
 
 mkdir -p "$layerPath"
-tar --no-recursion -rf "$layerPath/layer.tar" \
+
+# make sure /nix and /nix/store appear first in the archive
```
Comment (nlewo) [inline pkgs/build-support/docker/store-path-to-layer.sh]: Could you coment we are doing this to avoid file changes during the tar process.

**Review by nlewo — CHANGES_REQUESTED** [review]
Could you reformat your commit message?
Otherwise, the code looks good to me ;)

**Comment by purefn** [comment]
> Could you reformat your commit message?

Reformat it how? It's formatted according to the CONTRIBUTING.md unless there is a detail I'm missing.

**Comment by jonringer** [comment]
>Reformat it how? It's formatted according to the CONTRIBUTING.md unless there is a detail I'm missing.

Since dockerTools refers to an attr set, something like:
```
dockerTools.buildLayeredImage: fix building images in parallel
```

File: pkgs/build-support/docker/store-path-to-layer.sh (hunk 1 above)
[the remark this answers is not in the document]
Comment (purefn) [reply pkgs/build-support/docker/store-path-to-layer.sh]: Done.

File: pkgs/build-support/docker/store-path-to-layer.sh (hunk 2 above)
Comment (purefn) [reply pkgs/build-support/docker/store-path-to-layer.sh]: Done.

**Comment by purefn** [comment]
@nlewo Changed the final image tar to ignore the file changes it detected, added comments, and reformatted git commit as per @jonringer's suggestion. Should be good to merge now. Let me know if there is anything else.

File: pkgs/build-support/docker/default.nix [hunk 3]
```diff
@@ -625,7 +624,22 @@ rec {
           -i "$imageName" > image/repositories
 
         echo "Cooking the image..."
-        tar -C image --dereference --hard-dereference --sort=name --mtime="@$SOURCE_DATE_EPOCH" --owner=0 --group=0  --mode=a-w --xform s:'^./':: -c . | pigz -nT > $out
+        # tar exits with an exit code of 1 if files changed while it was
+        # reading them. it considers a change in the number of hard links
```
Comment (nlewo) [inline pkgs/build-support/docker/default.nix]:
```suggestion
        # reading them. It considers a change in the number of hard links
```

File: pkgs/build-support/docker/default.nix [hunk 4]
```diff
@@ -625,7 +624,22 @@ rec {
...
+        # to be a "change", which can cause this to fail if images are being
+        # built concurrently and auto-optimise-store is turned on. since
```
Comment (nlewo) [inline pkgs/build-support/docker/default.nix]:
```suggestion
        # built concurrently and the auto-optimise-store nix option is turned on. Since
```

File: pkgs/build-support/docker/default.nix [hunk 5]
```diff
@@ -625,7 +624,22 @@ rec {
...
+        # know the contents of these files will not change, we can reasonably
```
Comment (nlewo) [inline pkgs/build-support/docker/default.nix]:
```suggestion
        # the contents of these files will not change, we can reasonably
```

File: pkgs/build-support/docker/store-path-to-layer.sh [hunk 6]
```diff
@@ -5,16 +5,43 @@ set -eu
...
+shift
+
 layerPath="./layers/$layerNumber"
-echo "Creating layer #$layerNumber for $@"
+echo "Creating layer #$layerNumber for $storePath"
 
 mkdir -p "$layerPath"
-tar --no-recursion -rf "$layerPath/layer.tar" \
+
+# make sure /nix and /nix/store appear first in the archive.
```
Comment (nlewo) [inline pkgs/build-support/docker/store-path-to-layer.sh]:
```suggestion
# Make sure /nix and /nix/store appear first in the archive.
```

File: pkgs/build-support/docker/store-path-to-layer.sh [hunk 7]
```diff
@@ -5,16 +5,43 @@ set -eu
...
+# we create the directories here and use them because
```
Comment (nlewo) [inline pkgs/build-support/docker/store-path-to-layer.sh]:
```suggestion
# We create the directories here and use them because
```

File: pkgs/build-support/docker/store-path-to-layer.sh [hunk 8]
```diff
@@ -5,16 +5,43 @@ set -eu
...
+mkdir -p nix/store
+tar -cf "$layerPath/layer.tar"  \
     --mtime="@$SOURCE_DATE_EPOCH" \
-    --owner=0 --group=0 /nix /nix/store
-tar -rpf "$layerPath/layer.tar" --hard-dereference --sort=name \
+    --owner=0 --group=0 \
+    --transform='s,nix,/nix,' \
+    nix
+
+# we change into the /nix/store in order to avoid a similar
```
Comment (nlewo) [inline pkgs/build-support/docker/store-path-to-layer.sh]:
```suggestion
# We change into the /nix/store in order to avoid a similar
```

File: pkgs/build-support/docker/store-path-to-layer.sh [hunk 9]
```diff
@@ -5,16 +5,43 @@ set -eu
...
+# "file changed as we read it" error as above. Namely,
+# if we use the absolute path of /nix/store/123-pkg
+# and something new it added to the nix store while tar
```
Comment (nlewo) [inline pkgs/build-support/docker/store-path-to-layer.sh]:
```suggestion
# and something new is added to the nix store while tar
```

File: pkgs/build-support/docker/store-path-to-layer.sh [hunk 10]
```diff
@@ -5,16 +5,43 @@ set -eu
...
+# is running, it will detect a change to /nix/store and
+# fail. Instead, if we cd into the nix store and copy
+# the relative nix store path, tar will ignore changes
+# to /nix/store. In order to create the correct structure
+# in the tar file, we transform the relative nix store
+# path to the absolute store path
```
Comment (nlewo) [inline pkgs/build-support/docker/store-path-to-layer.sh]:
```suggestion
# path to the absolute store path.
```

**Review by nlewo — CHANGES_REQUESTED** [review]
Otherwise, LGTM! And thanks for all of these comments:)

**Comment by nlewo** [comment]
@GrahamcOfBorg test docker-tools

**Comment by nlewo** [comment]
Note I applied comment fixes in da261e36316b5b32009e8aa936e32df4b91e11bd.

Thanks!

**Comment by rimmington** [comment]
@nlewo Would it be possible to backport this fix to 19.09?

**Comment by nlewo** [comment]
@rimmington I'm sorry, I don't have time to backport it. Note also this is actually not a trivial backport: we need to backport another patch (because this one is buggy) and there is a merge conflict.

**Comment by rimmington** [comment]
@nlewo Ah, I didn't realise the backport is non-trivial; I may just wait for 20.03 then. Thanks for checking!
